openstf/feature_engineering/feature_applicator.py

```diff
@@ -0,0 +1,128 @@
+# SPDX-FileCopyrightText: 2017-2021 Alliander N.V. <korte.termijn.prognoses@alliander.com> # noqa E501>
+#
+# SPDX-License-Identifier: MPL-2.0
+from abc import ABC, abstractmethod
+
+import numpy as np
+import pandas as pd
+
+from openstf.feature_engineering.apply_features import apply_features
+from openstf.feature_engineering.general import (
+    add_missing_feature_columns,
+    remove_extra_feature_columns,
+)
+
+LATENCY_CONFIG = {"APX": 24}  # A specific latency is part of a specific feature.
+
+
+class AbstractFeatureApplicator(ABC):
+    def __init__(self, horizons: list, feature_set_list: list = None) -> None:
+        """
+
+        Args:
+            horizons: (list) list of horizons
+            feature_set_list: (list) List of requested features
+        """
+        if type(horizons) is not list and not None:
+            raise ValueError("Horizons must be added as a list")
+
+        self.feature_set_list = feature_set_list
+        self.horizons = horizons
+
+    @abstractmethod
+    def add_features(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Adds features to an input DataFrame
+
+        Args:
+            df: pd.DataFrame with input data to which the features have to be added
+        """
+        pass
+
+
+class TrainFeatureApplicator(AbstractFeatureApplicator):
+    def add_features(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Adds features to an input DataFrame.
+        This method is implemented specifically for a model train pipeline.
+        For larger horzions data is invalidated as when they are not available.
+        For example:
+            For horzion 24 hours the feature T-720min is not added as the load 720 minutes ago is not available 24 hours in advance.
+            In case of a horizon 0.25 hours this feature is added as in this case the feature is available.
+
+        Args:
+            df: pd.DataFrame with input data to which the features have to be added
+
+        Returns:
+            pd.DataFrame: Input DataFrame with an extra column for every added feature.
+
+        """
+
+        # Set default horizons if none are provided
+        if self.horizons is None:
+            self.horizons = [0.25, 24]
+
+        # Pre define output vairables
+        result = pd.DataFrame()
+        cols = []
+
+        # Loop over horizons and add corresponding features
+        for horizon in self.horizons:
+            res = apply_features(df.copy(), horizon=horizon)
+            res["Horizon"] = horizon
+            if len(res.columns) > len(cols):
```

Comment:
@JanMaartenvanDoorn  this part is unclear to me? Why is keeping track of this required?

Reply:
For larger horizons some features are not generated at all, this would result in a dataframe with fewer columns than for other horizons. This construction makes sure we return all columns in the end.

Reply:
I'm sure there is a good reason to do so. My point is that I should not need the explanation you just gave to understand the code. The code should be self explanatory by using function/variable naming, docstring, comments etc.